Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
```

Comment:
What is the behavior if the startup hook cannot be resolved, eg., DOTNET_STARTUP_HOOKS has an invalid path, file does not exist, is not an assembly, or does not contain Initialize(). Can we log something useful to the console? Ie not something like `Failed to initialize CoreCLR` which means we have to debug it.

Reply:
I added some details to he doc about the error behavior of the current implementation (https://github.com/dotnet/coreclr/pull/19486, with tests here: https://github.com/dotnet/core-setup/pull/4465). I'm not sure to what extent these exceptions should be called out in the design vs an implementation detail, but I've described them explicitly for now.

Reply:
Whether and how the user receives good error messages is a design consideration IMO

Reply:
Ok, I'll keep the documented exceptions but remove the error text from the spec.